### ISSUE TITLE
Question about discrete data

### ISSUE THREAD

**Issue by baderrasheed** [Trusted-AI/adversarial-robustness-toolbox#486]
Does anyone know how to implement attacks and defenses on discrete data where the cost function is not continuous and we can not make derivatives directly?

**Comment by beat-buesser** [comment]
Hi @baderrasheed Thank you very much for using ART!

I think you are asking an interesting question. To facilitate the discussion, could you please share more information about your dataset and the definition of your noncontinuous cost function? Maybe with a short example?

**Comment by baderrasheed** [comment]
Hi. I'm currently a PhD student and I'm working on adversarial attacks on
deep neural networks for security applications where the data is discrete
almost all the time. My current interest is applying adversarial attacks on
malware detectors. In such applications, the loss function is not
differentiable directly with respect to the input, and ART does not work
directly as far as I know.

**Comment by beat-buesser** [comment]
@baderrasheed I think if you are using a (common) neural network for classification you should be able to obtain gradients with respect to its input. As far as I understand based on your explanation the question is more how to apply these gradients. For example you could binarize your categorical features and apply `ProjectedGradientDescent` with appropriate arguments `norm=np.inf`, `eps=1.0` and `eps_step=1.0`  and estimator `clip_values=(0, 1)` which in combination would leave all features either 0 or 1 while minimising (or maximising) your loss function.
